### PR TITLE
Introduce staking

### LINKED ISSUE #373: Introduce staking action

### PR DESCRIPTION
**Pull request by ipdae**
- resolve #373 , #379 , #380
- see also 
  - https://www.notion.so/planetarium/210416-4871cd52802e41b58851896dc07c6d05
  - https://www.notion.so/planetarium/200223-7e9bbe7ef608483abc81b86560b282bc

### CODE REVIEW

**Comment by ipdae** [comment]
Merge after v100041

File: Lib9c/Action/CancelStaking.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            {
+                return states
+                    .SetState(stakingAddress, MarkChanged)
+                    .MarkBalanceChanged(GoldCurrencyMock, stakingAddress, context.Signer);
+            }
+
+            AgentState agentState = states.GetAgentState(context.Signer);
+            if (agentState is null)
+            {
+                throw new FailedLoadStateException("Aborted as the agent state is failed to load.");
```
Comment (area363) [inline Lib9c/Action/CancelStaking.cs:34]:
```suggestion
                throw new FailedLoadStateException("Aborted as the agent state failed to load.");
```

File: Lib9c/Action/CancelStaking.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            }
+
+            if (!states.TryGetState(stakingAddress, out Dictionary stateDict))
+            {
+                throw new FailedLoadStateException($"Aborted as the staking state is failed to load.");
```
Comment (area363) [inline Lib9c/Action/CancelStaking.cs:39]:
```suggestion
                throw new FailedLoadStateException($"Aborted as the staking state failed to load.");
```

File: Lib9c/Action/CancelStaking.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            StakingSheet stakingSheet = states.GetSheet<StakingSheet>();
+            int currentLevel = stakingState.Level;
+            if (currentLevel <= level)
+            {
+                throw new InvalidLevelException($"The level must be less than {currentLevel}.");
+            }
+
+            if (stakingState.End)
+            {
+                throw new StakingExpiredException($"{stakingAddress} is already expired on {stakingState.ExpiredBlockIndex}");
```
Comment (area363) [inline Lib9c/Action/CancelStaking.cs:54]:
```suggestion
                throw new StakingExpiredException($"{stakingAddress} has already expired on {stakingState.ExpiredBlockIndex}");
```

File: Lib9c/Action/Staking.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    .SetState(context.Signer, MarkChanged)
+                    .MarkBalanceChanged(GoldCurrencyMock, context.Signer, stakingAddress);
+            }
+
+            StakingSheet stakingSheet = states.GetSheet<StakingSheet>();
+
+            AgentState agentState = states.GetAgentState(context.Signer);
+            if (agentState is null)
+            {
+                throw new FailedLoadStateException("Aborted as the agent state is failed to load.");
```
Comment (area363) [inline Lib9c/Action/Staking.cs:37]:
```suggestion
                throw new FailedLoadStateException("Aborted as the agent state failed to load.");
```

File: Lib9c/Action/Staking.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            StakingState stakingState;
+            int currentLevel = 1;
+            if (states.TryGetState(stakingAddress, out Dictionary stateDict))
+            {
+                stakingState = new StakingState(stateDict);
+
+                if (stakingState.ExpiredBlockIndex < context.BlockIndex)
+                {
+                    throw new StakingExpiredException(
+                        $"{stakingAddress} is already expired on {stakingState.ExpiredBlockIndex}");
```
Comment (area363) [inline Lib9c/Action/Staking.cs:64]:
```suggestion
                        $"{stakingAddress} has already expired on {stakingState.ExpiredBlockIndex}");
```

File: Lib9c/Action/StakingReward.cs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+            if (!states.TryGetState(stakingAddress, out Dictionary stateDict))
+            {
+                throw new FailedLoadStateException($"Aborted as the staking state is failed to load.");
+            }
+
+            StakingState stakingState = new StakingState(stateDict);
+            if (stakingState.End)
+            {
+                throw new StakingExpiredException($"{stakingAddress} is already expired on {stakingState.ExpiredBlockIndex}");
+            }
```
Comment (area363) [inline Lib9c/Action/StakingReward.cs:48]:
```suggestion
            if (!states.TryGetAgentAvatarStates(context.Signer, avatarAddress, out AgentState agentState, out AvatarState avatarState))
            {
                throw new FailedLoadStateException($"Aborted as the avatar state of the signer failed to load.");
            }

            if (!states.TryGetState(stakingAddress, out Dictionary stateDict))
            {
                throw new FailedLoadStateException($"Aborted as the staking state failed to load.");
            }

            StakingState stakingState = new StakingState(stateDict);
            if (stakingState.End)
            {
                throw new StakingExpiredException($"{stakingAddress} has already expired on {stakingState.ExpiredBlockIndex}");
            }
```

**Review by area363 — APPROVED** [review]

File: Lib9c/Action/Staking.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+using Libplanet.Action;
+using Libplanet.Assets;
+using Nekoyume.Model.State;
+using Nekoyume.TableData;
+using static Lib9c.SerializeKeys;
+
+namespace Nekoyume.Action
+{
+    [Serializable]
+    [ActionType("staking")]
```
Comment (longfin) [inline Lib9c/Action/Staking.cs:15]: IMO `stake` is more natural because `staking` isn't a verb, but a noun.

File: Lib9c/Action/StakingReward.cs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+using Nekoyume.Model.Item;
+using Nekoyume.Model.Mail;
+using Nekoyume.Model.State;
+using Nekoyume.TableData;
+using static Lib9c.SerializeKeys;
+
+namespace Nekoyume.Action
+{
+    [Serializable]
+    [ActionType("staking_reward")]
```
Comment (longfin) [inline Lib9c/Action/StakingReward.cs:17]: Same way, I think `claim_staking_reward` is more natural for the action name.

File: Lib9c/Action/Staking.cs [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                    $"Expected staking round is {agentState.StakingRound}, but actual staking round is {stakingRound}.");
+            }
+
+            if (!stakingSheet.TryGetValue(level, out StakingSheet.Row _))
+            {
+                throw new SheetRowNotFoundException(nameof(StakingSheet), level);
+            }
+
+            Currency currency = states.GetGoldCurrency();
+            FungibleAssetValue requiredGold = currency * 0;
```
Comment (longfin) [inline Lib9c/Action/Staking.cs:52]: Is this zero intended? (if so, it would be better leaving comment for justifying)

Comment (ipdae) [reply Lib9c/Action/Staking.cs:52]: default value is intended. I will leave it as a comment.

**Review by longfin — APPROVED** [review]

**Pull request merged**